### PR TITLE
Raise `MismatchedABI` exceptions with detailed error messaging

### LINKED ISSUE #964: Improve error messages on mismatched argument types
### What was wrong?

When sending an argument with the wrong type to a contract function, people find it difficult to debug and end up frustrated, like in #625

### How can it be fixed?

We largely solve for the *general* case: that multiple functions can exist with the same name. Obviously we have to support that, but it's incredibly common to have names that are unique even without the argument types. In these cases, we can give clearer error messages about exactly which types are expected for which arguments.

For example, we could raise the function:
> ValidationError: a function with name testFunction is defined by the contract, but some argument types don't match:
> -  address _location :heavy_check_mark:
> -  uint256 _reading :heavy_check_mark:
> -  uint256 _minParameter :heavy_check_mark:
> -  uint256 _maxParameter :heavy_check_mark:
> -  bytes32 _codeHash :warning: This argument was supplied as a string that was too long. If you want to supply a hex string, this function only accepts values that are exactly 64 characters (plus 2 for 0x). Alternatively, supply this argument as python native bytes, with a length of 32 bytes. Actually supplied argument: "0xFFFFF..." (length: 67)

### PR DESCRIPTION
**Pull request by reedsa**
### What was wrong?

In `get_abi_element`, exception handling between functions and events is now shared. The messages reference `functions`, but should instead say `element` and provide additional details about any element that matches the name. The provided arguments and their types should be checked with the encodability for each element ABI in order to list each argument error.

Closes #964, #3482

### How was it fixed?

Add logic for mismatched ABI errors to invoke a new private method that generates a list of errors for each ABI that matches the provided `abi_element_identifier`. The messages returned depend on the number of matches and if there are arguments, the number of matching ABIs that have the same number of arguments is also considered.

The messages will include all potential matches (by name) so that users can identify the function or event they intended to reference. The specific messages for each match depend on the arguments provided.

When an argument is not encodable to a certain type, the message appends information about the argument position, the provided value, and the expected type.

### Todo:

- [X] Clean up commit history
- [X] Add or update documentation related to these changes
- [ ] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: tests/core/contracts/test_contract_method_to_argument_matching.py
Comment (kclowes) [inline tests/core/contracts/test_contract_method_to_argument_matching.py]: It took me a minute to figure out what this error message was telling me. Maybe something along the lines of: `Tried to find a matching ABI Element `a`, but encountered the following problems:` 🤷

File: tests/core/contracts/test_contract_method_to_argument_matching.py
Comment (kclowes) [inline tests/core/contracts/test_contract_method_to_argument_matching.py]: I'm not sure if you can make this wording work throughout, but maybe something like: `Multiple elements were found that accept 1 arguments`

File: tests/core/contracts/test_contract_method_to_argument_matching.py
Comment (kclowes) [reply tests/core/contracts/test_contract_method_to_argument_matching.py]: /cc our resident words guy ™️ @wolovim

**Review by kclowes — APPROVED** [review]
This is awesome! I left a few comments on the wording of the messages, but I think this will really help!

File: web3/utils/abi.py
Comment (fselmo) [inline web3/utils/abi.py]:
```suggestion
            f"{len(abi_element_input_types)} argument(s) but received {num_args} argument(s).\n"
```

File: web3/utils/abi.py
Comment (fselmo) [inline web3/utils/abi.py]:
```suggestion
        error += f"Multiple elements were found matching {num_args} argument(s).\n"
```

File: web3/utils/abi.py
Comment (fselmo) [inline web3/utils/abi.py]:
```suggestion
            f"{num_args} argument(s).\n"
```

File: web3/utils/abi.py
Comment (fselmo) [inline web3/utils/abi.py]:
```suggestion
            f"Found {num_abis_matching_arg_count} element(s) named `{name}` that takes "
```

File: web3/utils/abi.py
Comment (fselmo) [inline web3/utils/abi.py]:
```suggestion
            f"{num_args} argument(s).\n"
```

File: web3/utils/abi.py
Comment (fselmo) [inline web3/utils/abi.py]:
```suggestion
    error += f"Encountered problems with the following element(s) named `{name}`.\n"
```

File: tests/core/contracts/test_contract_method_to_argument_matching.py
Comment (fselmo) [reply tests/core/contracts/test_contract_method_to_argument_matching.py]: @kclowes I think it may be as simple as adding `(s)`. I made a comment in the actual messages under the abi utils. Do you think that clarifies well enough?

**Review by fselmo — COMMENTED** [review]
I made an initial pass to contribute to the messaging conversation. I think we can probably add `(s)` around the descriptors for dynamic singular or multiple values, e.g. `1 elements(s)`, `2 argument(s)`, etc...

I'll take a more careful pass at the rest on Monday 👀

File: web3/utils/abi.py
Comment (fselmo) [reply web3/utils/abi.py]: Is this always a singular element? If so, we can drop the `s` altogether.

File: tests/core/contracts/test_contract_method_to_argument_matching.py
Comment (kclowes) [reply tests/core/contracts/test_contract_method_to_argument_matching.py]: The word `matching` is what's confusing to me. To me, this says the element matches the argument (which I know is not the intent) but changing to to `accepts` clears it up. Either way, feel free to take or leave.

File: tests/core/utilities/test_event_interface.py
Comment (fselmo) [inline tests/core/utilities/test_event_interface.py:22]: Is this desirable? If we're under the `events` namespace, and for the sake of not breaking anything, we should probably keep the same exception. Thoughts there?

edit: We should also look at `ABIEventNotFound` cases.

File: web3/utils/abi.py
Comment (fselmo) [inline web3/utils/abi.py:702]: We should add a `with pytest.warns()` test for this and assert it doesn't warn when we use `get_abi_element()`. We've had a few issues where the warning wasn't quite working as expected recently.

File: tests/core/utilities/test_event_interface.py
Comment (fselmo) [reply tests/core/utilities/test_event_interface.py:22]: We should also check that `NoABIFunctionsFound` and `ABIFunctionNotFound` are still raised where they used to be raised if we want to go that route.

**Review by fselmo — COMMENTED** [review]
This looks great. I left some comments on error message wording. I think they could be good changes for clarity, including @kclowes comment.

I have some small outstanding questions on possible breaking changes as well. Exceptions that were raised before this PR and new classes seem to be raised now. Let's make sure we're not braking anything there.

File: web3/contract/contract.py
Comment (fselmo) [inline web3/contract/contract.py]: nit: this seems like a less clarifying change. We could keep `event_abi` here instead of the more generic "abi".

File: web3/utils/abi.py
Comment (reedsa) [reply web3/utils/abi.py]: Keeping plural here. The error lists one to many elements with the same name after this message.

File: tests/core/utilities/test_event_interface.py
Comment (reedsa) [reply tests/core/utilities/test_event_interface.py:22]: In the [`__getattribute__` function of the `ContractFunctions` class](https://github.com/ethereum/web3.py/blob/9baae32574e5dd655f3be50aee0f31c46548cc16/web3/contract/contract.py#L417) there is logic in raising exceptions that separates whether a contract was intialized with an `ABI` or the contract has an `ABI` but contains no functions. Functions have always raised `NoABIFound` when the ABI is not available.

Events now behave similar to functions, so the logic for the exceptions match what functions have. For functions, any attempt to use functions would raise the `NoABIFound` or `NoABIFunctionsFound` exception.

I agree we should not have breaking changes for this. Maybe one day this could be addressed as a breaking change. I can create a new issue to track that.

For this, I will change the logic for events so that `NoABIEventsFound` is raised and it wont be breaking. Would it be better to combine the logic so the exception and message is the same for both cases?

Alternatively, the same exception could be raised with different messages based on either the ABI not being present vs. the contract having no events. Something like:

```python
        if super().__getattribute__("abi") is None:
            raise NoABIEventsFound(
                "There is no ABI found for this contract.",
            )
        if "_events" not in self.__dict__:
            raise NoABIEventsFound(
                "The abi for this contract contains no event definitions. ",
                "Are you sure you provided the correct contract abi?",
            )
```

Comment (fselmo) [reply tests/core/utilities/test_event_interface.py:22]: Going to mark this resolved. We cleared this up that a contract with no ABI at all should indeed raise the `NoABIFound` exception. This test case seems to be expecting the wrong exception here so this is technically a bugfix.

File: tests/core/contracts/test_contract_class_construction.py
Comment (fselmo) [inline tests/core/contracts/test_contract_class_construction.py:91]: ❤️

**Review by fselmo — DISMISSED** [review]
~~lgtm!~~ I retracted the accept for now... I think we need to solve for some edge cases in messaging, etc. See comments below.

File: web3/utils/abi.py
Comment (fselmo) [inline web3/utils/abi.py:332]: @reedsa I approved but I think we need to tighten up the messaging a little bit. It looks like we only append problems when there are matching names. Perhaps we should make this code block a bit more like:

```py
if abi_signatures_matching_names:
    error += f"Tried to find a matching ABI element named `{name}`, but encountered the following problems:\n"

    for abi_signature in abi_signatures_matching_names:
        error += _build_abi_input_error(...)
```

Because if I run something like the following:

```bash
$ python -c "from web3.utils import get_abi_element;get_abi_element({}, 'nonexistent')"
```

I will get "encountered the following problems:" with no "problems" after it, just a blank.

**Comment by fselmo** [comment]
@reedsa I'm also seeing a recursion issue with simplistic ABI structures. If I try to `get_abi_element()` with an ABI that has a simple function and event with the same name, I enter into `_build_abi_input_error()` which tries to `get_abi_element()` with a signature and re-enters `_build_abi_input_error()`, etc...

reproduce with:

```bash
python -c "from web3.utils import get_abi_element;get_abi_element([{'type': 'function', 'name': 'Nonexistent'},{'type': 'event', 'name': 'Nonexistent'}], 'Nonexistent')"
```

**Comment by reedsa** [comment]
@fselmo Interesting to see the function/event name overlap. I think the general rule of thumb in solidity is using TitleCase for events, camelCase for functions. Though that is not explicitly enforced by Solidity as far as I know.

I'll have a look at the empty messages and recursive issue, good finds!

**Comment by fselmo** [comment]
> @fselmo Interesting to see the function/event name overlap. I think the general rule of thumb in solidity is using TitleCase for events, camelCase for functions. Though that is not explicitly enforced by Solidity as far as I know.

Yeah that was more of a lazy quick example to test. I think the important takeaway is a graceful handling of the recursion / edge case rather than just letting it recurse.

**Comment by reedsa** [comment]
@fselmo Found a better approach that eliminates recursive calls entirely. Also doing validations up front in `get_abi_element` by calling `validate_abi` and makes those type of errors much clearer.

**Comment by fselmo** [comment]
> @fselmo Found a better approach that eliminates recursive calls entirely. Also doing validations up front in `get_abi_element` by calling `validate_abi` and makes those type of errors much clearer.

I think we can probably still trim this messaging a bit. It seems quite lengthy when I think we can convey the same message more briefly. I'm seeing double errors as well, one for each of the identifying elements. Was that intentional?

```shell
$ python -c "from web3.utils import get_abi_element;get_abi_element([{'type': 'event', 'name': 'MyElement'}, {'type': 'function', 'name': 'MyElement'}], 'MyElement')"

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/fselmo/codigo/e/py/web3.py/web3/utils/abi.py", line 611, in get_abi_element
    raise MismatchedABI(error_diagnosis)
web3.exceptions.MismatchedABI: 
ABI Not Found!
Multiple elements were found that accept 0 argument(s).
Provided argument types: ()
Provided keyword argument types: {}

Tried to find a matching ABI element named `MyElement`, but encountered the following problems:
MyElement()
The provided identifier matches multiple elements.
If you meant to call `MyElement()`, please specify the full signature.
MyElement()
The provided identifier matches multiple elements.
If you meant to call `MyElement()`, please specify the full signature.
```

Doubled part:

```
MyElement()
The provided identifier matches multiple elements.
If you meant to call `MyElement()`, please specify the full signature.
```

What do you think? I can consider this a nit on my part if we think this is desirable. I'm not sure how much of an overhead it would be to not repeat the messaging per item found at the end.

**Comment by reedsa** [comment]
Oh that does look strange for that case, since it lists both ABI elements in the abi that match the name. Maybe the error message should include the type of each `function` or `event` to make that clear. Would make it even more verbose so may not be much better. Something like:

```sh
MyElement() - Type: `function`
The provided identifier matches multiple elements.
If you meant to call `MyElement()`, please specify the full signature.
MyElement() - Type: `event`
The provided identifier matches multiple elements.
If you meant to call `MyElement()`, please specify the full signature.
```

**Comment by fselmo** [comment]
@reedsa I think I see the point of the repeat (in this case). I'm also breaking the API on purpose... an ABI should never really define an event or function like that. I think this looks good. Messaging seems pretty clear to me and I'm not sure I mind the verbosity as much. It is at least very descriptive of the issues.

> Maybe the error message should include the type of each function or event to make that clear. 

I think including the type would be a great clarifying addition 👌🏼

**Comment by fselmo** [comment]
@reedsa it would be ideal if we could avoid repeating the obvious parts. Something more along the lines of:

```
The provided identifier matches multiple elements. If you meant to call `MyElement`, please specify the full signature. 
    - signature: "MyElement()", type: "event"
    - signature: "MyElement()", type: "function"
```

Or something along those lines. Specifically removing the unnecessary repeat of "The provided identifier matches multiple elements. If you meant to call `MyElement()`, please specify the full signature. "

**Review by fselmo — APPROVED** [review]
Good stuff. There's a lot of surface area here, maybe one more approval couldn't hurt but lgtm!

File: tests/core/contracts/test_contract_ambiguous_functions.py
Comment (kclowes) [inline tests/core/contracts/test_contract_ambiguous_functions.py:115]: This looks like it's exactly the same as lines 81-85?

File: tests/core/contracts/test_contract_class_construction.py
Comment (kclowes) [inline tests/core/contracts/test_contract_class_construction.py]: nit:
```suggestion
    # Contract `abi`` function should not override `abi` attribute
```

File: tests/core/contracts/test_contract_class_construction.py
Comment (kclowes) [inline tests/core/contracts/test_contract_class_construction.py]: It might also be worth adding a few of these tests on the ContractCaller class. I can't remember how the factory is instantiated over there. The test would look something like: 
```
assert contract.caller.abi() is True
```

File: tests/core/contracts/test_contract_class_construction.py
Comment (kclowes) [inline tests/core/contracts/test_contract_class_construction.py]: nit:
```suggestion
    # Contract `abi` function should not override `abi` attribute
```

File: tests/core/contracts/test_contract_class_construction.py
Comment (kclowes) [inline tests/core/contracts/test_contract_class_construction.py]: Feel free to take or leave, but we've been moving towards having separate tests instead of having `if` branches within the test. For example, having a test like `test_appropriate_exceptions_based_on_namespaces_with_no_abi` or something and moving those tests to a new test function. I can also see keeping them here since it's the same assertion. I'll leave it up to you :)

File: web3/_utils/contracts.py
Comment (kclowes) [inline web3/_utils/contracts.py:54]: ~~Are these being used anywhere now? We may be able to get rid of them entirely~~ For some reason Github won't let me delete. I see they're being used later on. Don't mind me :)

File: web3/utils/abi.py
Comment (kclowes) [inline web3/utils/abi.py]: Might be worth putting here that the first match gets returned. Is it the first match based on the ABI order?

**Review by kclowes — COMMENTED** [review]
Looking good! I didn't see anything major really. Just a few small things I noticed, mostly in the testing department: 
- I didn't comment on all of the instances I saw, but left a comment around splitting out branching `if` statements in tests into separate tests. Feel free to take or leave. 
- I also left a few nits and see a few warnings in the tests that weren't there before that could be cleaned up 🧹 (UserWarnings in the extracting_event_data test, and some DeprecationWarnings in the test_abi test)

File: tests/core/contracts/test_contract_class_construction.py
Comment (reedsa) [reply tests/core/contracts/test_contract_class_construction.py]: Yikes, this is really not working. I'm trying to figure out where the collisions are occurring. It seems to be intertwined with the `ContractFunctions` class.

Comment (reedsa) [reply tests/core/contracts/test_contract_class_construction.py]: @kclowes in my debugging, it appears that the `abi` property gets overridden when there is an `abi()` function in the contract. I assume this is because the special characters are ignored, causing both properties to be set with the function as the value. It should instead set `abi` to the ABI, and `abi()` to the function. It doesnt seem like there is a way to force the value to be the actual ABI. From what I see, setting the function to the `abi()` property also sets `abi`.

It is wildly confusing to trace this edge case through the code and perhaps doesnt make sense anyway. It is also an issue with the other class properties like `w3`/`w3()` and `address`/`address()` cases.

I am thinking it might just be better to disallow these specific function names to be used in a contract altogether. Would that be reasonable, especially since I was trying to do this on top of the other overloaded contract functions?

Comment (kclowes) [reply tests/core/contracts/test_contract_class_construction.py]: Is this just in the `ContractCaller` class? If so, think it would be fine to disallow functions with the `w3`/`abi`/`address` names via the `ContractCaller`, as long as it gets documented well. I wonder if we could raise an error or something if someone tries to access the function name? I don't necessarily think that we should block a contract abi with those method names in there, but I could be convinced if you feel strongly. If we do go that path (not allowing w3/abi/address function names on ContractCaller), would you add an issue so that we can track it?

Comment (reedsa) [reply tests/core/contracts/test_contract_class_construction.py]: I'm going to timebox this to see if there is anything I can isolate with these issues. I agree these names may not be widely used so it may be better to resolve in a later issue. I'll follow up with a new issue if it cannot be resolved in the next day or two.

Comment (reedsa) [reply tests/core/contracts/test_contract_class_construction.py]: @kclowes I am tinkering with the idea of using an underscore prefix for the class attributes. `_w3` / `_abi` / `_address` would be the new names. This is working at least to avoid collisions with `w3` / `abi` / `address`. If the contract contains the underscored name, the variable itself is no longer overwritten but the code does not access the function when called. It instead uses the class variable itself and throws an exception.

A couple questions:
1. Is changing these to private variables considered breaking?
2. Is it acceptable to fail initializing the contract if a contract uses this convention with the `_` prefix with the same name?

Comment (kclowes) [reply tests/core/contracts/test_contract_class_construction.py]: > Is changing these to private variables considered breaking?

As long as they're private, it shouldn't be breaking. A good smoke test would be to see what tests break. I can't remember if those really are private variables or not, but am happy to dig in further if you want a second pair of eyes.

> Is it acceptable to fail initializing the contract if a contract uses this convention with the _ prefix with the same name?

I think that's okay here. Looking into this more, `abi` and `address` are reserved in Solidity, so I don't think that we'll run into a problem with those two, and I think `w3` is enough of an edge case that we probably don't need to worry too much about it. Hopefully someone would raise an issue if they ran into problems.

**Comment by reedsa** [comment]
@kclowes I noticed that `UserWarning` has occurred for a while now but it's hard to pinpoint exactly what happened. There have been changes in late v6 and into v7 that may be the culprit. However I reverted a couple of those without success. I have created an [issue](https://github.com/ethereum/web3.py/issues/3521) to resolve that warning at some point in the near future.

File: web3/_utils/contract_sources/contract_data/event_contracts.py
Comment (kclowes) [inline web3/_utils/contract_sources/contract_data/event_contracts.py]: We should maybe recompile here with `v0.8.28` before merge, assuming that's the latest version

**Review by kclowes — APPROVED** [review]
LGTM! I left one comment about recompiling with solidity 0.8.28 but then I think it's ready when you're ready to pull the trigger!

**Comment by jurekovac** [comment]
This changes to async_contract.py and contract.py (especially (Async)ContractFunction.__call__) slows down the performance of obtaining the call of contract functions e.g. contract.function_name(*args) considerably. ~ 5x. This is a problem if having a lot of such calls.

**Pull request merged**
